### ISSUE TITLE
LocalBuildInfo should change at build time when components are selected

### ISSUE THREAD

**Issue by enolan** [haskell/cabal#2910]
You can specify which components to build like this: `runhaskell Setup.hs build my-testsuite`. But since `LocalBuildInfo` is generated at configure time, the `componentsConfigs` field list every component that was enabled then. This means you can't check whether certain components are being built.

I ran into this trying to update my PR #2844. `cabal-install`'s `Setup.hs` builds the manpage in a `postBuild` hook. If we're haven't built `cabal` yet and `dist/build/cabal` doesn't exist, then the `openFile` call to write `dist/build/cabal/cabal.1` will fail. We need to only try to make the manpage when building the executable component but it's impossible to find out at this stage.

**Comment by ezyang** [comment]
Hello @enolan, do you have suggested interface for being able to check if components are being built?

**Comment by enolan** [comment]
Hi Edward. My plan was to move the call to `readBuildTargets` from `Distribution.Simple.Build.build` to `Distribution.Simple.buildAction`. Inside the `preBuild` `buildHook` and `postBuild` hooks, the `componentsConfigs` field of `LocalBuildInfo` would only include components that were actually being built. I can't think of a reason the build phase would need to know about components that are configured but not actually being built. If anyone can, we'd need a new field in the lbi. I wrote this for cabal-install's Setup.hs:

```haskell
buildingExecutable :: LocalBuildInfo -> Bool
buildingExecutable lbi = any isCabalExe $ componentsConfigs lbi
  where isCabalExe (cname, _, _) = cname == CExeName "cabal"
```

**Comment by ezyang** [comment]
OK, some thoughts:

1. If you really want something only to be run per if something else is being built, this suggests we should have a per-component build hook, which gets called for every component being built (even better, you can get a ComponentLocalBuildInfo!) (ditto with postBuild/preBuild, I guess).

2. I don't think your plan works: componentsInBuildOrder will throw a hissy fit because it expects that executable you "just" wanted to build to depend on a component that is in the `componentsConfigs`, and it's not.

**Comment by enolan** [comment]
Yes, that makes sense. So we should either change the hooks to take an additional argument of `ComponentLocalBuildInfo` or add a new set of hooks with that parameter that default to doing nothing?

**Comment by ezyang** [comment]
Changing the hooks will break existing hooks in two ways: (1) they won't typecheck anymore, and (2) they'll get run multiple times but most users expect their hooks to run once. I'd add some new hooks.
